File: Minimum Path sum in a grid/Tabulation.py
def minimumPath(arr,m,n,dp):
        
        for i in range(m+1):
              
            for j in range(n+1):
                if i == 0 and j == 0: dp[i][j] = arr[i][j]
                else:
                    top = dp[i-1][j] if i > 0 else float('inf')
                    left = dp[i][j-1]  if j > 0 else float('inf')

                    dp[i][j] = arr[i][j] + min(top,left)
        
        return dp[m][n]

if __name__ == '__main__':
    arr = [
        [1,2,3],
        [2,5,1]
    ]

    m = len(arr)
    n = len(arr[0])

    dp = [ [-1]* n] * m

    print(minimumPath(arr,m-1,n-1,dp))

File: Minimum Path sum in a grid/Tabulation.java


public class Tabulation {
    public static void main(String[] args) {
        int[][] arr = {
            {1,2,3},
            {2,5,1}
        };

        int m = arr.length;
        int n = arr[0].length;

        int[][] dp = new int[m][n];

        for(int i = 0 ; i < m ;i++) {

            for(int j = 0 ; j < n; j++) {
                if(i == 0 && j == 0) dp[i][j] = arr[i][j];
                else {
                    int top  = Integer.MAX_VALUE / 2;
                    int left  = Integer.MAX_VALUE / 2;
                    if( i > 0) top = dp[i-1][j];
                    if(j > 0) left = dp[i][j-1];

                    dp[i][j] = arr[i][j] +  Math.min(top,left); 
                }

            }
        }

        System.out.println(dp[m-1][n-1]);

    }
}